core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala

```diff
@@ -48,83 +52,94 @@ private[spark] trait CachedReceiver {
+                                 nAndP.partitionId.toString,
+                                 EventPosition.fromSequenceNumber(requestSeqNo).convert,
+                                 DefaultEpoch,
+                                 receiverOptions),
+      "CachedReceiver creation."
+    )
+    _receiver.onComplete {
+      case Success(r) => r.setPrefetchCount(DefaultPrefetchCount)
+      case _          =>
     }
   }
 
-  private var _receiver: PartitionReceiver = _
-  private def receiver: PartitionReceiver = {
-    if (_receiver == null) {
-      throw new IllegalStateException(s"No receiver for $nAndP")
-    }
-    _receiver
+  private def closeReceiver(): Unit = {
+    val r = Await.result(receiver, InternalOperationTimeout)
+    _receiver = null
+    retryJava(r.close(), "CachedReceiver close.")
   }
 
-  private def errWrongSeqNo(requestSeqNo: SequenceNumber, receivedSeqNo: SequenceNumber): String =
-    s"requestSeqNo $requestSeqNo does not match the received sequence number $receivedSeqNo"
-
-  private def receive(requestSeqNo: SequenceNumber, batchSize: Int): EventData = {
-    def receiveOneEvent: EventData = {
-      var event: EventData = null
-      var i: java.lang.Iterable[EventData] = null
-      while (i == null) {
-        i = try {
-          receiver.receiveSync(1)
-        } catch {
-          case r: ReceiverDisconnectedException =>
-            throw new Exception(
-              "You are likely running multiple Spark jobs with the same consumer group. " +
-                "For each Spark job, please create and use a unique consumer group to avoid this issue.",
-              r
-            )
-        }
+  private def receiveOne: Future[Iterable[EventData]] = {
+    receiver
+      .flatMap { r =>
+        retryNotNull(r.receive(1), "CachedReceiver receive call")
       }
-      event = i.iterator.next
-      event
-    }
+      .map { e =>
+        e.asScala
+      }
+  }
 
-    var event: EventData = receiveOneEvent
-    if (requestSeqNo != event.getSystemProperties.getSequenceNumber) {
+  private def checkCursor(requestSeqNo: SequenceNumber): Future[Iterable[EventData]] = {
+    val event = Await.result(receiveOne, InternalOperationTimeout)
+    if (requestSeqNo != event.head.getSystemProperties.getSequenceNumber) {
       logWarning(
-        s"$requestSeqNo did not match ${event.getSystemProperties.getSequenceNumber}." +
+        s"$requestSeqNo did not match ${event.head.getSystemProperties.getSequenceNumber}." +
           s"Recreating receiver for $nAndP")
       closeReceiver()
       createReceiver(requestSeqNo)
-      event = receiveOneEvent
-      assert(requestSeqNo == event.getSystemProperties.getSequenceNumber,
-             errWrongSeqNo(requestSeqNo, event.getSystemProperties.getSequenceNumber))
+      receiveOne
+    } else {
+      Future { event }
+    }
+  }
+
+  private def receive(requestSeqNo: SequenceNumber, batchSize: Int): Iterator[EventData] = {
+    val first = checkCursor(requestSeqNo)
```

Comment:
Not sure if we need all this future logic here. Also doubting if it would make it more performant.

Reply:
See "the problem" above. There's a slight chance that the cursor can be in the wrong spot!